src/main/java/org/thoughtcrime/securesms/service/FetchForegroundService.java

```diff
@@ -0,0 +1,79 @@
+package org.thoughtcrime.securesms.service;
+
+import android.app.Notification;
+import android.app.Service;
+import android.content.Context;
+import android.content.Intent;
+import android.os.IBinder;
+import android.util.Log;
+
+import androidx.annotation.Nullable;
+import androidx.core.app.NotificationCompat;
+import androidx.core.content.ContextCompat;
+
+import org.thoughtcrime.securesms.ApplicationContext;
+import org.thoughtcrime.securesms.R;
+import org.thoughtcrime.securesms.notifications.FcmReceiveService;
+import org.thoughtcrime.securesms.notifications.NotificationCenter;
+import org.thoughtcrime.securesms.util.Util;
+
+public final class FetchForegroundService extends Service {
+  private static final String TAG = FcmReceiveService.class.getSimpleName();
+  private static final Object SERVICE_LOCK = new Object();
+  private static Intent service;
+
+  public static void start(Context context) {
+    GenericForegroundService.createFgNotificationChannel(context);
+    synchronized (SERVICE_LOCK) {
+      if (service == null) {
+        service = new Intent(context, FetchForegroundService.class);
+        ContextCompat.startForegroundService(context, service);
+      }
+    }
+  }
+
+  public static void stop(Context context) {
+    synchronized (SERVICE_LOCK) {
+      if (service != null) {
+        context.stopService(service);
+        service = null;
+      }
+    }
+  }
+
+  @Override
+  public void onCreate() {
+    Log.i(TAG, "Creating fetch service");
+    super.onCreate();
+
+    Notification notification = new NotificationCompat.Builder(this, NotificationCenter.CH_GENERIC)
+      .setContentTitle(getString(R.string.connectivity_updating))
+      .setSmallIcon(R.drawable.notification_permanent)
+      .build();
+
+    startForeground(NotificationCenter.ID_FETCH, notification);
+
+    // Start explicit fetch only after we marked ourselves as requiring foreground;
+    // this may help we on getting network and time adequately
+    // Fetch is started in background to not block the UI.
+    // We then run not longer than the max. of 20 seconds,
+    // see https://firebase.google.com/docs/cloud-messaging/android/receive .
+    Util.runOnAnyBackgroundThread(() -> {
+      Log.i(TAG, "Starting fetch");
+      if (!ApplicationContext.dcAccounts.backgroundFetch(19)) {
+        FetchForegroundService.stop(this);
+      }
```

Comment:
Do I understand correctly that we only stop the service if backgroundFetch() returns false, i.e. `dc_accounts_background_fetch()` returns 0, i.e. `accounts.is_null() || timeout_in_seconds <= 2`?

If so, why? Shouldn't we stop it in any case?

Reply:
see [dc_accounts_background_fetch()](https://c.delta.chat/classdc__accounts__t.html#a4b735a1e51222a06aa992773e42d06b0):
- usually,  `dc_accounts_background_fetch()` emits the event  `DC_EVENT_ACCOUNTS_BACKGROUND_FETCH_DONE` and returns true; `FetchForegroundService.stop(this)` os called as a reaction to this event 
- only if the event is not emitted (errors or timeout probably), `false` is returned. in this case, we stop the service here

Reply:
Ooooh ok. Maybe put a comment similar to: "Else, we need to wait for `DC_EVENT_ACCOUNTS_BACKGROUND_FETCH_DONE` to make sure all events are processed before calling `FetchForegroundService.stop(this)` - otherwise, the app may be terminated just after `backgroundFetch()`, leaving `DC_EVENT_INCOMING_MSG` unprocessed, resulting in no notification being shown."

FTR, slightly off-topic:

The only reason for `dc_accounts_background_fetch()` to return 0 is `accounts.is_null() || timeout_in_seconds <= 2`:

https://github.com/deltachat/deltachat-core-rust/blob/774307241163a959505f20725e486b2df6abaeb9/deltachat-ffi/src/lib.rs#L4880-L4898

In case of other errors or a timeout, the event is emitted and 1 is returned.

Reply:
i added a little pointer in the comment